experimental/devices/hd44780/hd44780.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2018 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// Package hd44780 controls the Hitachi LCD display chipset HD-44780
+//
+// Datasheet
+//
+// https://www.sparkfun.com/datasheets/LCD/HD44780.pdf
+package hd44780
+
+import (
+	"fmt"
+	"time"
+
+	"periph.io/x/periph/conn"
+	"periph.io/x/periph/conn/gpio"
+)
+
+// lineTwo offset for the second line in the LCD buffer.
+const lineTwo = 0x40
+
+// Dev is the 4-bit addressing device for HD-44780
+type Dev struct {
+	// data pins
+	dataPins []gpio.PinOut
+
+	// register select pin
+	rsPin gpio.PinOut
+
+	// enable pin
+	enablePin gpio.PinOut
+}
+
+// New creates and initializes the LCD device
+//	data - references to data pins
+//	rs - rs pin
+//	e - strobe pin
+func New(data []gpio.PinOut, rs, e gpio.PinOut) (*Dev, error) {
+	if len(data) != 4 {
+		return nil, fmt.Errorf("expected 4 data pins, passed %s", len(data))
+	}
+	dev := &Dev{
+		dataPins:  data,
+		enablePin: e,
+		rsPin:     rs,
+	}
+	if err := dev.Init(); err != nil {
+		return nil, err
+	}
+	return dev, nil
+}
+
+// Init initializes the HC-44780 chipset, clears the screen buffer and moves cursor to the
+// home of screen (line 0, column 0).
+func (r *Dev) Init() error {
+	if err := r.clearBits(); err != nil {
+		return err
+	}
+
+	delayMs(15)
+
+	if err := r.rsPin.Out(gpio.Low); err != nil {
+		return err
+	}
+	if err := r.enablePin.Out(gpio.Low); err != nil {
+		return err
+	}
+
+	if err := r.bulkSendData(resetSequence, r.write4Bits); err != nil {
+		return err
+	}
+
+	if err := r.bulkSendData(initSequence, r.writeInstruction); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Halt will re-initialize the display.
+func (r *Dev) Halt() error {
+	return r.Init()
+}
+
+func (r *Dev) String() string {
+	return "HD44870, 4 bit mode"
+}
+
+// Cls clears the LCD screen
+func (r *Dev) Cls() error {
+	if err := r.writeInstruction(0x01); err != nil {
+		return err
+	}
+	delayMs(2)
+	return nil
+}
+
+// SetCursor positions the cursor
+//	line - screen line, 0-based
+//	column - column, 0-based
+func (r *Dev) SetCursor(line uint8, column uint8) error {
+	if err := r.writeInstruction(0x80 | (line*lineTwo + column)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Print the data string
+//	data string to display
+func (r *Dev) Print(data string) error {
+	for _, v := range []byte(data) {
+		if err := r.WriteChar(v); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// PrintAt the data string at specified position
```

Comment:
I don't see the need of exporting SetCursor(), Print() and WriteChar() with this function.

I'd recommend to explicitly return an error when the string overflows the known boundaries.

Reply:
If the cursor is out of range - nothing will happen (at least with the current device I have for testing). So I presume that the invalid addresses are ignored.

Reply:
Please document this.